## Likelihood Ratio (LR) and Likelihood Ratio Test (LRT)

# Intuition

- **Likelihood Ratio (LR):** The ratio of the likelihoods of two models given the observed data, measuring relative support for one model over the other.  
- **Bayes Factor (BF):** The ratio of the marginal likelihoods of two models, integrating over their parameters with prior distributions to assess relative evidence.  
- **Likelihood Ratio Test (LRT):** A hypothesis test comparing two nested models using the likelihood ratio, with the test statistic following a chi-squared distribution.


**FIGURE PLACEHOLDER:** ![Likelihood Intuition Cartoon](image_placeholder)

# Notations

## Likelihood Ratio (LR)
The **Likelihood Ratio (LR)** compares two fully-specified models by taking the ratio of their likelihoods given the observed data $x$:

$$
LR(M_2, M_1) := \frac{L(M_2)}{L(M_1)} = \frac{p(x|M_2)}{p(x|M_1)}
$$

Large values of $LR(M_2, M_1)$ indicate that the data are much more probable under **Model 2** than under **Model 1**, thus supporting **Model 2**. On the other hand, small values of the likelihood ratio suggest stronger support for **Model 1**. 


## Likelihood Ratio Test (LRT)

The Likelihood Ratio Test (LRT) is a statistical method used to compare two models by using the likelihood ratio. Often the likelihood-ratio test statistic is expressed as a difference between the log-likelihoods

$$
LRT(M_2, M_1) := - 2 \log \left( \frac{L(M_1)}{L(M_2)} \right) = - 2 \left[ \log L(M_1) - \log L(M_2) \right]
$$

Under certain regularity conditions, the LRT statistic follows a chi-squared distribution with degrees of freedom equal to the difference in the number of parameters between $M_2$ and $M_1$. It is commonly used in frequentist hypothesis testing to assess whether a more complex model provides a significantly better fit than a simpler, nested model.

# Example

Here we continue the example used in the `6_1_likelihood.ipynb` notebook.

In [26]:
set.seed(2)

# Simulate true mean and effect size
baseline <- 170  # Population mean of the trait (e.g., height in cm) when the genetic variant has no effect (Model 1)
theta_true <- 2  # True effect size of the genetic variant. This represents the change in height (in cm) associated with each additional minor allele (Model 2)
sd_y <- 1  # Standard deviation of the trait (e.g., variability in height measurement within the population)

# Simulate genotypes for three individuals
genotype <- c(1, 2, 0)

# Simulate height values for three individuals based on genotypes
n = length(genotype)
height_values <- rnorm(n, mean = baseline + theta_true * genotype, sd = sd_y)

# Display the simulated data
data.frame(genotype = genotype, height = height_values)


genotype height  
1 1        171.1031
2 2        174.1848
3 0        171.5878

In [27]:
# Define the likelihood function for a normal distribution
likelihood <- function(theta, sd, y) {
  prod(dnorm(y, mean = theta, sd = sd))
}

# Likelihood for Model 1: No genetic effect
likelihood_1_multi_samples <- likelihood(theta = baseline, sd = sd_y, height_values)

# Likelihood for Model 2: Genetic effect
likelihood_2_multi_samples <- likelihood(theta = baseline + theta_true * genotype, sd = sd_y, height_values)

# Print the likelihoods for both models
cat("Likelihood for Model 1:", likelihood_1_multi_samples, "\n")
cat("Likelihood for Model 2:", likelihood_2_multi_samples, "\n")

Likelihood for Model 1: 1.542167e-06 
Likelihood for Model 2: 0.01183428 


Now let's compute the **likelihood ratio** between $M_2$ and $M_1$:

In [28]:
# Compute the likelihood ratio (LR)
lr <- likelihood_2_multi_samples / likelihood_1_multi_samples

# Print the likelihood ratio
cat("Likelihood Ratio (LR) for Model 2 vs Model 1:", lr, "\n")

Likelihood Ratio (LR) for Model 2 vs Model 1: 7673.803 


Then we perform the **Likelihood Ratio Test (LRT)**,

In [29]:
# Compute the Likelihood Ratio Test (LRT) statistic
lrt_statistic <- -2 * (log(likelihood_1_multi_samples) - log(likelihood_2_multi_samples))

# Print the LRT statistic
cat("Likelihood Ratio Test Statistic (LRT):", lrt_statistic, "\n")

Likelihood Ratio Test Statistic (LRT): 17.89114 


**Interpreting the Results**
- Likelihood Ratio (LR): A larger LR suggests that $M_2$ (genetic effect) better fits the data than $M_1$ (no genetic effect).
- Likelihood Ratio Test (LRT): If the LRT statistic is large, we prefer $M_2$ over $M_1$. The test statistic follows a chi-square distribution with 1 degree of freedom.

# Summary
- The **"likelihood ratio"** that compares two fully-specified (discrete) models is simply the ratio of the likelihood for the two models given data $D$: 

$$
LR(M_2, M_1) := \frac{L(M_2)}{L(M_1)} = \frac{p(D|M_2)}{p(D|M_1)}
$$

where $L(M)$ denotes the likelihood for model $M$ under data $x$.

- The **"log-likelihood ratio"** is defined as the logarithm of the likelihood ratio:

$$
LLR(M_2, M_1) := - 2 \log \left( \frac{L(M_1)}{L(M_2)} \right) = - 2 \left[ \log L(M_1) - \log L(M_2) \right] = - 2 \left[ l(M_1) - l(M_2) \right]
$$

where $l(M_i)$ denotes the log-likelihood for model $M_i$ under data $x$. The log-likelihood ratio measures the relative support for model $M_2$ over model $M_1$ on a **logarithmic scale**.

- A natural question is,  **what value of the LR (or LLR) should we treat as "convincing" evidence for one model vs another**? The short answer here is that this is **context dependent**, which depends on the relative plausibility of the models we are comparing. For more discussions, refer to this [notes](https://stephens999.github.io/fiveMinuteStats/LR_and_BF.html) by Matthew Stephens.

# Supplementary Materials

## Example 1 -- allele frequency

Here we use the same example 1 in the `6_1_likelihood.ipynb` notebook, and we have calculated the likelihood for the two models $M_S$ and $M_F$.

### Calculate LR

In [8]:
x = c(1,0,1,0,0,1)

fS = c(0.40, 0.12,0.21,0.12,0.02,0.32)
fF = c(0.8,0.2,0.11,0.17,0.23,0.25)

L = function(f,x){ prod(f^x*(1-f)^(1-x)) }

In [9]:
L(fS,x)
L(fF,x)

[1] 0.02039955

[1] 0.01124816

The **likelihood ratio (LR)** is simply the ratio of those two values:

In [10]:
L(fS,x)/L(fF,x)

[1] 1.81359

### How to interpret LR?

> So $LR(M_S,M_F;x)$ is 1.8135904. This means that the data favor the tusk coming from a savanna elephant by a factor of about 1.8. This is a fairly modest factor – not large enough to draw a convincing conclusion. We will have more to say about interpreting LRs, and what values might be considered “convincing” later.
>
> Note that we have deliberately focused on the likelihood ratio, and not the actual likelihood values themselves. This is because actual likelihood values are generally not useful - it is only the ratios that matter when comparing the models. One way of thinking about this is that the actual likelihood values are very context dependent, and so likelihoods from different data sets are not comparable with one another. However, the meaning of the likelihood ratio is in some sense consistent across contexts: LR =1.8 means that the data favour the first model by a factor of 1.8 whatever the context.

## Example 2 -- concertration of protein in blood

Here we use the same example 2 in the `6_1_likelihood.ipynb` notebook, and we have calculated the likelihood for the two models $M_0$ (normal individual group) and $M_1$ (diseased individual group).

### Calculate LR

In [11]:
X_val=4.02
# dgamma(x, shape, rate = 1, scale = 1/rate, log = FALSE) returns the Density for the Gamma distribution with parameters shape and scale at x
y0_val = dgamma(X_val,scale=0.5,shape=2)
y1_val = dgamma(X_val,scale=1,shape=2)
y0_val
y1_val

[1] 0.005182728

[1] 0.07217092

In [12]:
y1_val/y0_val

[1] 13.92528

### Interpretation

The $LR(M_1,M_0;x)$ is 13.9, i.e., the data favours the individual being diseased ($M_1$) over being normal ($M_0$) by a factor of approximately 14.



# Recommended Reading

- Section *Likelihood Ratio and Likehood* in [FiveMinuteStats](http://stephens999.github.io/fiveMinuteStats/index.html) by Matthew Stephens
